packages/core/src/odata/v4/filter-function.ts

```diff
@@ -0,0 +1,60 @@
+/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
+
+import {
+  FilterFunctionParameterType,
+  BooleanFilterFunction,
+  NumberFilterFunction,
+  StringFilterFunction
+} from '../common/filter';
+import {
+  FilterFunctionReturnType,
+  createFilterFunction
+} from '../common/filter/filter-function';
+import { Entity } from './entity';
+import { DateFilterFunction } from './date-filter-function';
+
+export function filterFunction<EntityT extends Entity>(
+  functionName: string,
+  returnType: 'boolean' | 'bool',
+  ...parameters: FilterFunctionParameterType<EntityT>[]
+): BooleanFilterFunction<EntityT>;
+
+export function filterFunction<EntityT extends Entity>(
+  functionName: string,
+  returnType: 'int' | 'double' | 'decimal',
+  ...parameters: FilterFunctionParameterType<EntityT>[]
+): NumberFilterFunction<EntityT>;
+
+export function filterFunction<EntityT extends Entity>(
+  functionName: string,
+  returnType: 'string',
+  ...parameters: FilterFunctionParameterType<EntityT>[]
+): StringFilterFunction<EntityT>;
+
+export function filterFunction<EntityT extends Entity>(
+  functionName: string,
+  returnType: 'datetimeoffset',
+  ...parameters: FilterFunctionParameterType<EntityT>[]
+): DateFilterFunction<EntityT>;
+
```

Comment:
So much boiler plate here and in the filter-function of common. I see that you opened a question on StackOverflow. Let's hope that there is some elegant solution not using inheritance.